app/src/main/java/net/gsantner/markor/frontend/textview/LineNumbersTextView.java

```diff
@@ -0,0 +1,338 @@
+package net.gsantner.markor.frontend.textview;
+
+import android.content.Context;
+import android.graphics.Canvas;
+import android.graphics.Paint;
+import android.graphics.Rect;
+import android.text.Editable;
+import android.text.Layout;
+import android.text.TextWatcher;
+import android.util.AttributeSet;
+import android.view.View;
+import android.view.ViewTreeObserver;
+import android.widget.EditText;
+import android.widget.ScrollView;
+
+import androidx.appcompat.widget.AppCompatTextView;
+
+@SuppressWarnings("UnusedReturnValue")
+public class LineNumbersTextView extends AppCompatTextView {
+    private LineNumbersDrawer lineNumbersDrawer;
+    private boolean lineNumbersEnabled;
+    private ScrollView scrollView;
+    private final ViewTreeObserver.OnScrollChangedListener onScrollChangedListener = new ViewTreeObserver.OnScrollChangedListener() {
+        private long lastTime;
+
+        @Override
+        public void onScrollChanged() {
+            final long time = System.currentTimeMillis();
+            if (time - lastTime > 125) {
+                lastTime = time;
+                refresh();
+            }
+        }
+    };
+
+    public LineNumbersTextView(Context context) {
+        super(context);
+    }
+
+    public LineNumbersTextView(Context context, AttributeSet attrs) {
+        super(context, attrs);
+    }
+
+    public LineNumbersTextView(Context context, AttributeSet attrs, int defStyleAttr) {
+        super(context, attrs, defStyleAttr);
+    }
+
+    @Override
+    protected void onVisibilityChanged(View changedView, int visibility) {
+        super.onVisibilityChanged(changedView, visibility);
+        if (isLineNumbersEnabled() && visibility == View.VISIBLE) {
+            refresh();
+        }
+    }
+
+    @Override
+    protected void onDraw(Canvas canvas) {
+        super.onDraw(canvas);
+        if (lineNumbersEnabled) {
+            lineNumbersDrawer.draw(canvas);
+        }
+    }
+
+    public void refresh() {
+        setText(""); // To activate LineNumbersTextView refresh
+        if (getWidth() == 0) {
+            lineNumbersDrawer.getEditText().postInvalidate();
+        }
+    }
+
+    public void setup(final EditText editText, ScrollView scrollView) {
+        this.scrollView = scrollView;
+        lineNumbersDrawer = new LineNumbersDrawer(editText, this);
+    }
+
+    public void setLineNumbersEnabled(final boolean enabled) {
+        lineNumbersEnabled = enabled;
+        if (lineNumbersEnabled) {
+            lineNumbersDrawer.prepare();
+            if (scrollView != null) {
+                scrollView.getViewTreeObserver().addOnScrollChangedListener(onScrollChangedListener);
```

Comment:
I think you can do `getViewTreeObserver().addOnScrollChangedListener` on the EditText, so you don't have to get and pass _primaryScrollView from the DocumentEditAndViewFragment